Feature/openrpc (#3)
etclabscore/ECIPs: ECLIPs/ECLIP-2.md

> # Add OpenRPC Service Discovery

Reviewer: ```suggestion
# Add OpenRPC Service Discovery To JSON-RPC Services
```

Reviewer 2: JSON-RPC

Reviewer 2: Or if we follow [StarIPs](https://github.com/BelfordZ/starIPs), it would be `ECIP-07be0c83`

    ECLIP: 2

>     Title: Add OpenRPC discover method

Author: ```suggestion
    Title: Add OpenRPC Service Discovery
```

>     Title: Add OpenRPC Service Discovery

Reviewer 3: ...to multi-geth specifically? or to multigeth/geth/parity

Author: trying to keep it pretty generic so it could be proposed anywhere, even as say a BIP

Reviewer 4: or for anything that has a JSONRPC server, eh?

Author: yea exactly

>     Title: Add OpenRPC Service Discovery

Reviewer: ```suggestion
    Title: Add OpenRPC Service Discovery To JSON-RPC Services
```

Reviewer 2: JSON-RPC

Reviewer: edited

    Author: Shane Jonas, Zachary Belford
    Status: WIP
    Type: Interface
    Created: 2019-02-25

# Abstract
### What is this?

> This is a proposal to add [OpenRPC](https://github.com/open-rpc/spec) support by adding the method [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) to the projects [JSON-RPC](https://www.jsonrpc.org/specification) API to enable automation and tooling.

Reviewer: So I think this needs to be fleshed out a little moar. What projects (or kinds of projects) is this specifically geared toward? Have you got a list of example projects?

Reviewer: ```suggestion
This is a proposal to add [OpenRPC](https://github.com/open-rpc/spec) support to existing and future Ethereum-order RPC services by adding the method [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) to the projects [JSON-RPC](https://www.jsonrpc.org/specification) APIs, enabling automation and tooling.
```

Author: `to these projects's`

should maybe read
`to this projects`, `this`, + dont need the `'s`

Reviewer: ... to add OpenRPC support to [...] services by adding the method `rpc.discover` to this projects APIs, enabling...
?

# Motivation
### What is the problem?

- Maintenence time and accuracy on:
    - documentation
    - APIs
    - clients
- Not all JSON-RPC servers return the same methods or return the same version of the API.
- There is nothing that provides both human *and* machine-readable interfaces to be able to interact with JSON-RPC. 
- Communication between services in different programming languages is not easy.
- Implementation usually comes at the cost of user experience consuming the API.

# Implementation
### How do I Solve the problem?
JSON-RPC APIs can support the OpenRPC specification by implementing a service discovery method that will return the [OpenRPC document](https://github.com/open-rpc/spec#openrpc-document) for the JSON-RPC API. The method MUST be named `rpc.discover`. The `rpc.` prefix is a reserved method prefix for [JSON-RPC 2.0 Specification](https://www.jsonrpc.org/specification) system extensions.

# Specification

### What is OpenRPC?

> The [OpenRPC](https://github.com/open-rpc/spec) Specification defines a standard, programming language-agnostic interface description for [JSON-RPC 2.0](https://www.jsonrpc.org/specification) APIs, which allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic. When properly defined via OpenRPC, a consumer can understand and interact with the remote service with a minimal amount of implementation logic. Similar to what interface descriptions have done for lower-level programming, the OpenRPC Specification removes guesswork in calling a service.

Reviewer: ```suggestion
The [OpenRPC](https://github.com/open-rpc/spec) Specification defines a standard, programming language-agnostic interface description for [JSON-RPC 2.0](https://www.jsonrpc.org/specification) APIs, which allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic. When properly defined via OpenRPC, a consumer can understand and interact with the remote service with a minimal amount of implementation logic, and share these logic patterns across use cases. Similar to what interface descriptions have done for lower-level programming, the OpenRPC Specification removes guesswork in calling a service.
```

# Rationale
### Why would we do this?

> Services need to figure out how to talk to each other. If we really want to build the next generation automation then having up to date libraries, documented APIs, and modern tool are going to provide easy discovery, on-boarding and interaction.

Reviewer: ```suggestion
Services need to figure out how to talk to each other. If we really want to build the next generation of automation, then having up to date libraries, documented APIs, and modern tools are going to provide easy discovery, on-boarding, and enable end user and developer interaction.
```

> Services need to figure out how to talk to each other. If we really want to build the next generation automation, then having up to date libraries and docingumented APIs and modern tool are going to provide easy discovery, onboarding and interaction.

Author: ```suggestion
Services need to figure out how to talk to each other. If we really want to build the next generation automation then having up to date libraries, documented APIs, and modern tool are going to provide easy discovery, on-boarding and interaction.
```

Use cases for machine-readable [JSON-RPC 2.0](https://www.jsonrpc.org/specification) API definition documents include, but are not limited to:

- A common vocabulary and document will keep developers, testers, architects, and technical writers all in sync.
- Server stubs/skeletons generated in many languages
- Clients generated in many languages 
- Mock Server generated in many languages 

> - Tests generated many language tests

Author: ```suggestion
- Tests generated in many languages
```

- Documentation Generation

> 

Reviewer 5: ```suggestion

### Possible benefits of standardized blockchain methods
Since Blockchains share a lot of common functionality, it would make sense for them to expose the same set of RPC methods that are common to all blockchains e.g. `getBlock(height)` or `getTransaction(txHash)`. The reuse of such blockchain methods could make certain tasks easier for services that need to integrate with multiple blockchains. Examples of services that could benefit from this include, but are not limited to:
- blockchain explorer for multiple chains
- exchanges listing new coins

```

Author: I think this is awesome. but maybe is another Improvement Proposal around BTC/ETC and other chains coming together on a good naming scheme for methods for these APIs?

Or I have other ideas around doing a meta-API t hat would allow this but its out of the scope of this ECLIP.

Reviewer 5: @shanejonas agree, this seems more like something on top of this proposal

# Alternative

> Since [OpenRPC](https://github.com/open-rpc/spec) documents just describe a [JSON-RPC](https://www.jsonrpc.org/specification) APIs services and are represented in JSON format. These documents may either be produced and served statically or be generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) is implemented.

Author: ```suggestion
Since [OpenRPC](https://github.com/open-rpc/spec) documents just describe a [JSON-RPC](https://www.jsonrpc.org/specification) APIs services and are represented in JSON format. These documents may either be produced and served statically or be generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) is implemented.
```

> Since [OpenRPC](https://github.com/open-rpc/spec) documents just describe a [JSON-RPC](https://www.jsonrpc.org/specification) APIs services and are represented in JSON format. These documents may either be produced and served statically or be generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) is implemented.

Author: ```suggestion
Since [OpenRPC](https://github.com/open-rpc/spec) documents just describe a [JSON-RPC](https://www.jsonrpc.org/specification) APIs services and are represented in JSON format. These documents may either be produced and served statically or be generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method is implemented.
```

> Since [OpenRPC](https://github.com/open-rpc/spec) documents just describe a [JSON-RPC](https://www.jsonrpc.org/specification) APIs services and are represented in JSON format. These documents may either be produced and served statically or be generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method is implemented.

Reviewer 4: ```suggestion
[OpenRPC](https://github.com/open-rpc/spec) documents just describe [JSON-RPC](https://www.jsonrpc.org/specification) APIs services, and are represented in JSON format. These documents may be produced and served statically OR generated dynamically from an application and returned via the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method. This gives projects and communities the flexibility to adopt tools before the [`rpc.discover`](https://github.com/open-rpc/spec#service-discovery-method) method is implemented.
```

# Add OpenRPC Service Discovery

>     ECLIP: 2

Author: ```suggestion
    ECLIP: ?
```
